### ISSUE TITLE
Compute forces on the IPU. 

### ISSUE THREAD

**Issue by AlexanderMath** [graphcore-research/pyscf-ipu#66]
nanoDFT computes forces on the CPU using `def grad(..)` on [line 230](https://github.com/graphcore-research/pyscf-ipu/blob/3eaeaebf7bdf29cd592f87db2c2210c7d0025d68/nanoDFT/nanoDFT.py#L267). To run this on the IPU we could port [lines 269-273](https://github.com/graphcore-research/pyscf-ipu/blob/3eaeaebf7bdf29cd592f87db2c2210c7d0025d68/nanoDFT/nanoDFT.py#L269) and [line 283](https://github.com/graphcore-research/pyscf-ipu/blob/3eaeaebf7bdf29cd592f87db2c2210c7d0025d68/nanoDFT/nanoDFT.py#L283). 

Different strategies: 
1. Compile libcint to poplar and replace all `mol.intor(..)` with corresponding poplar calls (ERI is only real problem). 
2. Use Jax implementation from [D4FT](https://github.com/sail-sg/d4ft/tree/main/d4ft/integral) for the forward pass of the `mol.intor(..)` and match up the derivative with the calls on lines 270-273 ([pyscfad](https://github.com/fishjojo/pyscfad) did this on CPU so their code may be helpful).  
3. Line 230 uses [this theorem](https://faculty.washington.edu/miller/51816/Feynman.pdf) to compute gradients. We could instead use `jax.grad(..)`. For that we need to fix all calls in `_nanoDFT(..)` which don't support derivatives.

**Comment by AlexanderMath** [comment]
**Note:** If we use autodiff/`jax.grad` we need to store "activations" during `jax.lax.fori_loop(0, opts.its, nanoDFT_iteration, ..)`. We might use [jax.checkpoint](https://jax.readthedocs.io/en/latest/_autosummary/jax.checkpoint.html) to only store `density_matrix` of shape `(N,N)` in each iteration, and then during backprop keep all activations within an iteration in memory. This leads to a peak memory consumption of `N^2*num_iterations + floats_within_one_iteration`.  

**Note:** It takes 3x more memory to store `ERI_grad = mol.intor("int2e_ip")` because `ERI_grad.shape=(N, N, N, N, 3)`. `ERI_grad` is only used twice see lines [244-245](https://github.com/graphcore-research/pyscf-ipu/blob/3eaeaebf7bdf29cd592f87db2c2210c7d0025d68/nanoDFT/nanoDFT.py#L244) (similar einsum to how `ERI` is used). 
```
vj = - jnp.einsum('sijkl,lk->sij', eri, dm0) # (3, N, N)
vk = - jnp.einsum('sijkl,jk->sil', eri, dm0) # (3, N, N)
 ```
Since we are only using once we may as-well compute sequentially.

**Comment by AlexanderMath** [comment]
Another implementation (thanks to @paolot-gc) https://github.com/erikkjellgren/SlowQuant/blob/master/slowquant/molecularintegrals/electronrepulsion.py

**Comment by AlexanderMath** [comment]
Another implementation (watch our for license) https://github.com/theochem/gbasis
